File: leetcode-scala/src/main/scala/com/leetcode/sequences/ProductOfArrayExceptSelf.scala
package com.leetcode.sequences

object ProductOfArrayExceptSelf {

  def productExceptSelf(nums: Array[Int]): Array[Int] = {
    val length = nums.length
    val result = Array.ofDim[Int](length)
    var leftProduct = 1

    result(length - 1) = nums.last

    for (i <- (length - 2) until 0 by -1) {
      result(i) = nums(i) * result(i + 1)
    }

    for (i <- 0 until length - 1) {
      result(i) = leftProduct * result(i + 1)
      leftProduct *= nums(i)
    }

    result(length - 1) = leftProduct

    result
  }
}

File: leetcode-java/src/main/java/com/melalex/leetcode/ProductOfArrayExceptSelf.java
package com.melalex.leetcode;

public class ProductOfArrayExceptSelf {

  public int[] productExceptSelf(int[] nums) {
    var result = new int[nums.length];

    var buff = 1;

    for (var i = 0; i < nums.length; i++) {
      result[i] = buff;
      buff *= nums[i];
    }

    buff = 1;

    for (var i = nums.length - 1; i >= 0; i--) {
      result[i] *= buff;
      buff *= nums[i];
    }

    return result;
  }
}
